examples/pytorch/graphsage/model.py

```diff
@@ -0,0 +1,150 @@
+                 n_hidden,
+                 n_classes,
+                 n_layers,
+                 activation,
+                 dropout,
+                 aggregator_type):
+        super().__init__(in_feats, n_hidden, n_classes, n_layers, activation, dropout, aggregator_type)
+        self.n_hidden = n_hidden
+        self.n_classes = n_classes
+        
+    def forward(self, blocks, x):
+        h = x
+        for l, (layer, block) in enumerate(zip(self.layers, blocks)):
+            h = layer(block, h)
+            if l != len(self.layers) - 1:
+                h = self.activation(h)
+                h = self.dropout(h)
+        return h
+
+    def inference(self, g, device, batch_size):
+        feat = g.ndata['feat']
+        sampler = dgl.dataloading.MultiLayerFullNeighborSampler(1, prefetch_node_feats=['feat'])
+        dataloader = dgl.dataloading.DataLoader(
+                g, torch.arange(g.num_nodes()).to(g.device), sampler, device=device,
+                batch_size=batch_size, shuffle=False, drop_last=False,
+                num_workers=0)
+
+        for l, layer in enumerate(self.layers):
+            y = torch.empty(
+                g.num_nodes(), self.n_hidden if l != len(self.layers) - 1 else self.n_classes,
+                device='cpu', pin_memory=True)
+            feat = feat.to(device)
+            for input_nodes, output_nodes, blocks in tqdm.tqdm(dataloader):
+                # use an explicitly contiguous slice
+                x = feat[input_nodes]
+                h = layer(blocks[0], x)
+                if l != len(self.layers) - 1:
+                    h = self.activation(h)
+                    h = self.dropout(h)
+                # be design, our output nodes are contiguous so we can take
+                # advantage of that here
+                y[output_nodes[0]:output_nodes[-1]+1] = h.to('cpu')
+            feat = y
+        return y
+
+class GraphSAGEBatchMultiGPU(GraphSAGEBatch):
+    def inference(self, g, device, batch_size):
+        """
+        Perform inference in layer-major order rather than batch-major order.
+        That is, infer the first layer for the entire graph, and store the
+        intermediate values h_0, before infering the second layer to generate
+        h_1. This is done for two reasons: 1) it limits the effect of node
+        degree on the amount of memory used as it only proccesses 1-hop
+        neighbors at a time, and 2) it reduces the total amount of computation
+        required as each node is only processed once per layer.
+
+        Parameters
+        ----------
+            g : DGLGraph
+                The graph to perform inference on.
+            device : context
+                The device this process should use for inference
+            batch_size : int
+                The number of items to collect in a batch.
+
+        Returns
+        -------
+            tensor
+                The predictions for all nodes in the graph.
+        """
```

Comment:
To be consistent with `GraphSAGEBatch`, either we do not add doc for it, or we have a brief doc for both.

Reply:
Well, I think it might be better to have such detailed comment in the code, especially for something that is not straightforward. (performing inference layer-major instead of batch-major is typical/standardized or well-known?) 

Also, I did notice another similar comment, like the `inference` function within `advanced/model.py` file. We should avoid repeating ourselves. I will keep in mind when I refactor advanced/ scripts.

I agree with you on consistency here. How about I move this detailed comment to GraphSAGEBatch:inference function, and only paste its first sentence to GraphSAGEBatchMultiGPU:inference?

Reply:
The proposal sounds good to me. However, I guess we can wait till we decide if they should be separate examples.